app/services/settings-v2/video.ts

```diff
@@ -3,80 +3,201 @@ import { Inject } from 'services/core/injector';
+        (key: keyof obs.IAdvancedStreaming | keyof obs.ISimpleStreaming) => {
+          this.SET_VIDEO_SETTING(key, data[key], display);
+          if (
+            invalidFps(this.contexts[display].video.fpsNum, this.contexts[display].video.fpsDen)
+          ) {
+            this.createDefaultFps(display);
+          }
+        },
+      );
     }
   }
 
-  establishVideoContext() {
-    if (this.state.videoContext) return;
+  establishVideoContext(display: TDisplayType = 'horizontal') {
+    if (this.contexts[display]) return;
+
+    this.contexts[display] = obs.VideoFactory.create();
+    this.state[display] = {} as obs.IVideoInfo;
+    this.migrateSettings(display);
+    this.contexts[display].video = this.state[display];
+
+    return !!this.contexts[display];
+  }
+
+  destroyVideoContext(display: TDisplayType = 'horizontal') {
+    if (this.contexts[display]) {
+      const context: obs.IVideo = this.contexts[display];
+      context.destroy();
+    }
+    this.DESTROY_VIDEO_CONTEXT(display);
 
-    this.SET_VIDEO_CONTEXT();
+    return !!this.contexts[display];
+  }
 
-    this.migrateSettings();
-    obs.VideoFactory.videoContext = this.state.videoContext;
+  createDefaultFps(display: TDisplayType = 'horizontal') {
+    this.setVideoSetting('fpsNum', 30, display);
+    this.setVideoSetting('fpsDen', 1, display);
   }
 
   @debounce(200)
-  updateObsSettings() {
-    obs.VideoFactory.videoContext = this.state.videoContext;
-    obs.VideoFactory.legacySettings = this.state.videoContext;
+  updateObsSettings(display: TDisplayType = 'horizontal') {
+    switch (display) {
+      case 'horizontal': {
+        this.contexts.horizontal.video = this.state.horizontal;
+        break;
+      }
+      case 'green': {
+        this.contexts.green.video = this.state.green;
+        break;
+      }
+      default: {
+        this.contexts.horizontal.video = this.state.horizontal;
+      }
+    }
+  }
+
+  resetToDefaultContext() {
+    for (const context in this.contexts) {
+      if (context !== 'horizontal') {
+        this.destroyVideoContext(context as TDisplayType);
+      }
+    }
+  }
+
+  setVideoSetting(key: string, value: unknown, display: TDisplayType = 'horizontal') {
+    this.SET_VIDEO_SETTING(key, value, display);
+
+    if (display === 'horizontal') {
+      this.updateObsSettings();
+    } else if (display === 'green') {
+      // if the display is green, also update the persisted settings
+      this.updateObsSettings('green');
+      this.greenService.setVideoSetting({ [key]: value });
+    }
+  }
+
+  shutdown() {
+    displays.forEach(display => {
+      const context = this.contexts[display];
+      if (context) {
+        context.destroy();
+      }
+    });
   }
 
-  setVideoSetting(key: string, value: unknown) {
-    this.SET_VIDEO_SETTING(key, value);
-    this.updateObsSettings();
+  @mutation()
+  DESTROY_VIDEO_CONTEXT(display: TDisplayType = 'horizontal') {
+    this.contexts[display] = null as obs.IVideo;
```

Comment:
usually we don't want side effects in out mutations, so this should probably just exist in `destroyVideoContext`

Reply:
Done